nltk/translate/bleu_score.py

```diff
@@ -140,10 +140,14 @@ def corpus_bleu(list_of_references, hypotheses, weights=(0.25, 0.25, 0.25, 0.25)
     :type weights: list(float)
     :param smoothing_function:
     :type smoothing_function: SmoothingFunction
-    :param auto_reweigh:
+    :param auto_reweigh: Option to re-normalize the weights uniformly.
     :type auto_reweigh: bool
-    :param emulate_multibleu: bool
+    :param emulate_multibleu: Option to emulate multi-bleu.pl behavior
+    :type emulate_multibleu: bool
+    :param backoff: Option to backoff to the largest n-gram overlap if there's less than one 4-gram overlap.
```

Comment:
Can you change this to "if set to `True` the precision for n-gram overlap with no overlap will evaluate to 1".

Reply:
The backoff doesn't always evaluate 1.0, (i) it evaluates to 1.0 only if the unigrams matches totally and there's no bigram matches or (ii) when the largest order of n-th ngram overlap is 1.0 and n < 4. 

Actually, we could also swap it to be `emulate_multibleu=True` and `backoff=False` by default but I'm afraid we have to handle the complaints we had before using the backoff strategy. 

I would really like to hear more people's view on which to set as default because there wasn't any response when the `emulate_multibleu` was introduced in #1554.

Reply:
I did not mention the BLEU score here, I mentioned the precision for the n-gram (or p_n in Papineni et. al.). The effect of using `backoff=True` is essentially matehmatically equivalent to letting p_n=1 in these case. Look through the equations carefully. 

It might help to use SmoothingFunction.method1, when you have set epsilon=1. Which (if we have `ref = [['hello', 'how', 'are', 'you']]` and `sent = ['hello', 'how', 'are']`) returns the same results as if SmoothingFunction.method0 was used. And if you have a look at the equation for smoothing 1 in http://acl2014.org/acl2014/W14-33/pdf/W14-3346.pdf I think you'll understand =)

In my mind either you use `emulate_multibleu=True` by default, or you specify clearly in the docs that the function intentionally does not follow the equations in the papers.

Reply:
Possibly, we need more inputs from users and other devs to know what should be the default `emulate_multibleu` or `backoff. 

Regarding the documentation, the new extended docstring  in method0 is already quite verbose =(

----

`backoff=True` effects don't always results `p_n=1`

In the simplest use case that most chatbot system might see, yes it does. Largely because the input is already `p_1=1`, `p_2=1` and `p_3=1`:

```
>>> from nltk.translate.bleu_score import SmoothingFunction
>>> from nltk.translate.bleu_score import modified_precision
>>> ref = ['hello', 'how', 'are', 'you']
>>> hyp = ['hello', 'how', 'are']
>>> zero = SmoothingFunction().method0

>>> [modified_precision([ref], hyp, n=i) for i in range(1,5)]
[Fraction(3, 3), Fraction(2, 2), Fraction(1, 1), Fraction(0, 1)]

>>> p_n = [modified_precision([ref], hyp, n=i) for i in range(1,5)]

>>> zero(p_n, emulate_multibleu=False, backoff=True)
[Fraction(3, 3), Fraction(2, 2), Fraction(1, 1)]
```

But in the case that the input with non-perfect lower order of ngram precision, that's the case for most MT outputs, the removal of the final order of n-gram doesn't result in `p_n=1` e.g.

```
>>> ref = ['hello', 'how', 'are', 'you', 'you']
>>> hyp = ['you', 'you', 'hello', 'how', 'are']
>>> p_n = [modified_precision([ref], hyp, n=i) for i in range(1,5)]
>>> p_n
[Fraction(5, 5), Fraction(3, 4), Fraction(1, 3), Fraction(0, 2)]
>>> zero(p_n, emulate_multibleu=False, backoff=True)
[Fraction(5, 5), Fraction(3, 4), Fraction(1, 3)]
```

Although to make it a real backoff function, there's more work to be done with regards to conditional probabilities but for now I think this is adequate.

Reply:
This would be off-topic, I think:

Regarding `epsilon` and lapacian smoothing, it's always weird to use laplace ;P 
It's added so that it follows Chen and Cherry (2014) method 1 to 6. The usual MT strategy is to go with `method3`.

And on the averaging sentence BLEU -> corpus metric, I think that's not the norm in MT but I see how it can be useful for localized scores per chat utterance as oppose to global score.

Reply:
@alvations 

Let's go through your last example in detail. For easier referencing later we can define the geometric mean equation: `BLEU = BP * exp(\sum_{n-1}^N w_n log(p_n))` as (1), for simplicity I'll assume w_n = 1/4 and I will disregard BP. 

Originally following their equations you have: `p_n = [5/5, 3/4, 1/3, 0/2]`, if you stick that into (1) you get: `exp(1/4 * log(1) + 1/4 * log(3/4) + 1/4 * log(1/3) + 1/4 * log(0/2))`. -> `exp(1/4 * 0 + 1/4 *(-0.288) + 1/4 * (-1.099) + 1/4 * -\infty)` -> `exp(-\infty)` -> `0`

With your backoff strategy you have: `p_n = [5/5, 3/4, 1/3, 0/2]` -> `[5/5, 3/4, 1/3]`, if you stick that into (1) you get: `exp(1/4 * log(1) + 1/4 * log(3/4) + 1/4 * log(1/3))`. -> `exp(1/4 * 0 + 1/4 *(-0.288) + 1/4 * (-1.099))` -> `exp( 1/4 *(-0.288) + 1/4 * (-1.099))`-> `exp(-0.346)` -> `0.7070`

I claim that this is equivalent of setting p_n=1 for the n:s that had no overlap (4 in this case). Let's try that: `p_n = [5/5, 3/4, 1/3, 0/2]` -> `[5/5, 3/4, 1/3, 1]`, if you stick that into (1) you get: `exp(1/4 * log(1) + 1/4 * log(3/4) + 1/4 * log(1/3) + 1/4 * log(1))`. -> `exp(1/4 * 0 + 1/4 *(-0.288) + 1/4 * (-1.099) + 1/4 * 0)` -> `exp( 1/4 *(-0.288) + 1/4 * (-1.099))` -> `exp(-0.346)` -> `0.7070`.

Do you follow? The main point is that when you set p_n=1 the log(p_n) will evaluate to 0. And you have a method that is equivalent to your backoff strategy.

So what you have done with your backoff strategy is to introduce a new type of smoothing. Which if following Chen and Cherry's notation could be described as `m_n' = l_n if m_n = 0`. 

And what buggs (sorry if I'm sounding condescending) me is that this smoothing technique is the worst I possibly can imagine. What you want from a smoothing technique (and what is the common theme through all smoothing techniques in presented by Chen and Cherry) is to if `m_n = 0` replace it with some **small** constant value compared to `l_n`. What you mathematically is doing is replacing it with `l_n` itself.

Reply:
Thank you for the clarification. Yes, essentially the backoff gives full credits for the ngrams that didn't exist:

```
>>> math.exp(sum([math.log(1.0), math.log(0.8), math.log(0.5)]))
0.4

>>> math.exp(sum([math.log(1.0), math.log(0.8), math.log(0.5), math.log(1.0)]))
0.4
```

I've removed it, please change the code for the doctest and unittest accordingly.

Pardon my bad math.